source/opt/loop_utils.cpp

```diff
@@ -0,0 +1,379 @@
+    if (incoming_phi) {
+      return *incoming_phi;
+    }
+
+    // Check if one of the loop exit basic block dominates |bb_id|.
+    for (const ir::BasicBlock* e_bb : exit_bb_) {
+      if (dom_tree_.Dominates(e_bb->id(), bb_id)) {
+        incoming_phi = bb_to_phi[e_bb->id()];
+        assert(incoming_phi && "No closing phi node ?");
+        return *incoming_phi;
+      }
+    }
+
+    // Process parents, they will returns their suitable phi.
+    // If they are all the same, this means this basic block is dominated by a
+    // common block, so we won't need to build a phi instruction.
+    std::vector<uint32_t> incomings;
+    for (uint32_t pred_id : cfg_->preds(bb_id)) {
+      incomings.push_back(GetOrBuildIncoming(pred_id).result_id());
+      incomings.push_back(pred_id);
+    }
+    uint32_t first_id = incomings.front();
+    size_t idx = 0;
+    for (; idx < incomings.size(); idx += 2)
+      if (first_id != incomings[idx]) break;
+
+    if (idx >= incomings.size()) {
+      incoming_phi = bb_to_phi[incomings[1]];
+      return *incoming_phi;
+    }
+
+    // We have at least 2 definitions to merge, so we need a phi instruction.
+    ir::BasicBlock* block = cfg_->block(bb_id);
+
+    opt::InstructionBuilder<> builder(context_, &*block->begin());
+    incoming_phi = builder.AddPhi(insn_type_, incomings);
+
+    rewritten.insert(incoming_phi);
+
+    return *incoming_phi;
+  }
+
+  ir::IRContext* context_;
+  ir::CFG* cfg_;
+  const opt::DominatorTree& dom_tree_;
+  uint32_t insn_type_;
+  std::unordered_map<uint32_t, ir::Instruction*> bb_to_phi;
+  std::unordered_set<ir::Instruction*> rewritten;
+  const std::unordered_set<ir::BasicBlock*>& exit_bb_;
+};
+}  // namespace
+
+void LoopUtils::CreateLoopDedicateExits() {
+  ir::Function* function = loop_->GetHeaderBlock()->GetParent();
+  ir::LoopDescriptor& loop_desc = *context_->GetLoopDescriptor(function);
+  ir::CFG& cfg = *context_->cfg();
+  opt::analysis::DefUseManager* def_use_mgr = context_->get_def_use_mgr();
+
+  constexpr ir::IRContext::Analysis PreservedAnalyses =
+      ir::IRContext::kAnalysisDefUse |
+      ir::IRContext::kAnalysisInstrToBlockMapping;
+
+  // Gathers the set of basic block that are not in this loop and have at least
+  // one predecessor in the loop and one not in the loop.
+  std::unordered_set<uint32_t> exit_bb_set;
+  loop_->GetExitBlocks(context_, &exit_bb_set);
+
+  std::unordered_set<ir::BasicBlock*> new_loop_exits;
+  bool made_change = false;
+  // For each block, we create a new one that gathers all branches from
+  // the loop and fall into the block.
+  for (uint32_t non_dedicate_id : exit_bb_set) {
+    ir::BasicBlock* non_dedicate = cfg.block(non_dedicate_id);
+    const std::vector<uint32_t>& bb_pred = cfg.preds(non_dedicate_id);
+    // Ignore the block if:
+    //   - all the predecessors are in the loop;
+    //   - and has an unconditional branch;
+    //   - and any other instructions are phi.
```

Comment:
You check that all other instructions are phis.

Reply:
Why does a dedicated exit have to contain only phi instructions?  This also contradicts the comment in the ".h" for this function that says structured control flow already has a dedicated exit.

At the same time, are you taking into consideration that merge block for the loop is not necessarily the exit block for the loop?  For example, 

```
for( int i = 0; i < 10; i++ ) {
// bb1
s += i;
}
// bb2
if (s > 10) //bb 3 s = 10;
// bb 4 selection merge
// bb 5 loop merge
...
```

In this case it is possible that the merge block for the OpLoopMerge in bb1 is bb5, not bb2.  I believe this is valid.  Does this effect the way things are being done?

Reply:
Hum very good point.

Specifically for your example: I don't think bb5 is a valid merge block (at least I find the choice strange, as bb2 post-dominate all exits), but I see your main point. I believe this code would generate a close structure (not quite the same though):
```
int i = 0;
while (i < 11) {
  s += i;
  if (i != 10) {
     if (s > 10) s = 10;
     break;
  }
  i++;
}
```
You have a similar pattern, but  would bb5 post-dominate all exits.

Generally, this function and the LCSSA transformation rely only on the capability to have access to the header block and the loop blocks. The update of the merge block will only happen if there is only 1 exit block (which is not correct if your example is valid).

Reply:
For the phi-instruction, I wasn't sure if we needed such a strong requirement. I just forgot to remove that part.

I added that code at some point because, I thought it would be easier to maintain the LCSSA if we have this property, but it is easy to do without.